x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java

```diff
@@ -0,0 +1,308 @@
+            autodetectExecutorService.submit(processor::process);
+        } catch (EsRejectedExecutionException e) {
+            // If submitting the operation to read the results from the process fails we need to close
+            // the process too, so that other submitted operations to threadpool are stopped.
+            try {
+                IOUtils.close(process);
+            } catch (IOException ioe) {
+                logger.error("Can't close autodetect", ioe);
+            }
+            onFinish.accept(e);
+            return;
+        }
+
+        StateStreamer stateStreamer = new StateStreamer(client);
+        Executor executor = new Executor(stateStreamer, processor, autodetectWorkerExecutor, process);
+        if (continueRunning.get() == false) {
+            onFinish.accept(null);
+            return;
+        }
+        executor.execute();
+    }
+
+    void setTaskToFailed(String reason, ActionListener<PersistentTask<?>> listener) {
+        SnapshotUpgradeTaskState taskState = new SnapshotUpgradeTaskState(
+            SnapshotUpgradeState.FAILED,
+            task.getAllocationId(),
+            reason);
+        task.updatePersistentTaskState(taskState, ActionListener.wrap(
+            listener::onResponse,
+            f -> {
+                logger.warn(
+                    () -> new ParameterizedMessage("[{}] [{}] failed to set task to failed", task.getJobId(), task.getSnapshotId()),
+                    f);
+                listener.onFailure(f);
+            }
+        ));
+    }
+
+    private class Executor {
+
+        private final StateStreamer stateStreamer;
+        private final JobSnapshotUpgraderResultProcessor processor;
+        private final ExecutorService autodetectWorkerExecutor;
+        private final AutodetectProcess process;
+
+        Executor(StateStreamer stateStreamer,
+                 JobSnapshotUpgraderResultProcessor processor,
+                 ExecutorService autodetectWorkerExecutor,
+                 AutodetectProcess process) {
+            this.stateStreamer = stateStreamer;
+            this.processor = processor;
+            this.autodetectWorkerExecutor = autodetectWorkerExecutor;
+            this.process = process;
+        }
+
+        void execute() {
+            this.restoreState();
+        }
+
+        protected final Map<String, Integer> outputFieldIndexes() {
+            Map<String, Integer> fieldIndexes = new HashMap<>();
+            // time field
+            fieldIndexes.put(job.getDataDescription().getTimeField(), 0);
+            int index = 1;
+            for (String field : job.getAnalysisConfig().analysisFields()) {
+                if (AnalysisConfig.ML_CATEGORY_FIELD.equals(field) == false) {
+                    fieldIndexes.put(field, index++);
+                }
+            }
+            fieldIndexes.put(LengthEncodedWriter.CONTROL_FIELD_NAME, index);
+            return fieldIndexes;
+        }
+
+        void writeHeader() throws IOException {
```

Comment:
It seemed weird to me that I had to write the header ahead of time as I only wanted to load and then persist the snapshot, but I ran into weird failures when I didn't. 

I can dig further if desired.

Reply:
I am not surprised you had to write a header.  You could probably get away with writing one with just the control field (field name `.`).  But it's not particularly important, so I'm happy to leave what's here.